### ISSUE TITLE
Generate input data for NN

### ISSUE THREAD

**Issue by geojamesc** [cropmapteam/Scotland-crop-map#42]
as needed for #39

**Comment by quantoidb** [comment]
Hi James, 
I am having a heck of a time reading TIF images into an R array :( Not giving up, but still looking into it. 
How difficult would it be to have the Kelso images in JPG format?

**Comment by geojamesc** [comment]
The images are geotiffs. I`m not sure what will happen to the data if it is converted to a jpeg.

In RStudio are you able to do something like this:

library(raster)

library(rgdal)

GDALinfo("S1B_20180630_30_asc_175813_175838_DV_Gamma0_GB_OSGB_RCTK_SpkRL_9_resized.tif")

img <- raster("S1B_20180630_30_asc_175813_175838_DV_Gamma0_GB_OSGB_RCTK_SpkRL_9_resized.tif")

summary(img)

This is me following

https://datacarpentry.org/r-raster-vector-geospatial/aio.html

**Comment by geojamesc** [comment]
I don`t think jpg format will support the output of the S1 radar data. When I try to convert one of the resized images from (geo)tiff to JPEG using:

james@dlib-rainier:~/Desktop$ gdal_translate -of JPEG S1B_20180630_30_asc_175813_175838_DV_Gamma-0_GB_OSGB_RCTK_SpkRL_9_resized.tif S1B_20180630_30_asc_175813_175838_DV_Gamma-0_GB_OSGB_RCTK_SpkRL_9_resized.jpg

I get:

Input file size is 64, 64
0ERROR 6: JPEG driver doesn't support 2 bands.  Must be 1 (grey), 3 (RGB) or 4 bands.

I suppose it might be possible to write out each band as a seperate JPEG but this is a further complication.

**Comment by quantoidb** [comment]
This is what I get when I look at a single image. 

> GDALinfo("nnTrain/ASSF/S1A_20180107_30_asc_175836_175901_DV_Gamma-0_GB_OSGB_RCTK_SpkRL_296_resized.tif") 
rows        64 
columns     64 
bands       2 
lower left origin.x        377828.4 
lower left origin.y        630317.4 
res.x       7.360938 
res.y       12.53875 
ysign       -1 
oblique.x   0 
oblique.y   0 
driver      GTiff 
projection  +proj=tmerc +lat_0=49 +lon_0=-2 +k=0.9996012717 +x_0=400000 +y_0=-100000 +ellps=airy
+towgs84=446.448,-125.157,542.06,0.15,0.247,0.842,-20.489 +units=m +no_defs 
file        nnTrain/ASSF/S1A_20180107_30_asc_175836_175901_DV_Gamma-0_GB_OSGB_RCTK_SpkRL_296_resized.tif 
apparent band summary:
   GDType hasNoDataValue NoDataValue blockSize1 blockSize2
1 Float32           TRUE         NaN         16         64
2 Float32           TRUE         NaN         16         64
apparent band statistics:
         Bmin       Bmax Bmean Bsd
1 -4294967295 4294967295    NA  NA
2 -4294967295 4294967295    NA  NA
Metadata:
AREA_OR_POINT=Area  

Warning messages:
1: In GDALinfo("nnTrain/ASSF/S1A_20180107_30_asc_175836_175901_DV_Gamma-0_GB_OSGB_RCTK_SpkRL_296_resized.tif") :
  statistics not supported by this driver
2: In GDALinfo("nnTrain/ASSF/S1A_20180107_30_asc_175836_175901_DV_Gamma-0_GB_OSGB_RCTK_SpkRL_296_resized.tif") :
  statistics not supported by this driver

**Comment by quantoidb** [comment]
Here is my R code. I am using Keras/TenserFlow functions (image_data_generator,  flow_images_from_directory) to read in the images. 

# Training data re-sizing (other options help augment photos) 
train_data_gen = image_data_generator(
        rescale = 1/255 #,
        #rotation_range = 40,
        #width_shift_range = 0.2,
        #height_shift_range = 0.2,
        #shear_range = 0.2,
        #zoom_range = 0.2,
        #horizontal_flip = TRUE,
        #fill_mode = "nearest" )

# Training images 
train_image_array_gen <- flow_images_from_directory(train_dir, 
                                                    generator = train_data_gen,
                                                    target_size = target_size,
                                                    color_mode="rgb",
                                                    class_mode = "categorical",
                                                    classes = crops_list,
                                                    batch_size = batch_size, 
                                                    seed = 42)

**Comment by geojamesc** [comment]
doing:

 b1 <- raster("/home/james/Desktop/S1.tif", band=1)
hist(b1)

The above seems to display a histogram in RStudio with what look like expected values from the supplied tiff file. So to me it looks like R is reading the S1 data from the geotiff file. BUT I don`t know enough (hardly anything) about R to tell much more than that.

I don`t know what flow_images_from_directory() is doing. But why color_mode="rgb". The S1 data is not RGB data. It`s 2 band data.

**Comment by quantoidb** [comment]
flow_images_from_directory() generates batches of data from images in a directory. 

COLOR_MODE available function options: "grayscale", "rbg". Default: "rgb". Whether the images will be converted to have 1 or 3 color channels. 

Does 2 band mean it is grayscale?

**Comment by quantoidb** [comment]
does CHANNEL and BAND mean the same thing?

**Comment by quantoidb** [comment]
Going back to basics. Here are a few screen-shots of me just looking at a few random images. Not what I expected. 

<img width="1438" alt="our_images_exploration-code" src="https://user-images.githubusercontent.com/10274858/61149471-2957f700-a4d9-11e9-9b25-43a3d457f6b3.png">
![our_image1](https://user-images.githubusercontent.com/10274858/61149472-2957f700-a4d9-11e9-9850-2b4ac7a53eb8.png)
![our_image2](https://user-images.githubusercontent.com/10274858/61149473-29f08d80-a4d9-11e9-9eec-1eda87d6f50c.png)
![our_image3](https://user-images.githubusercontent.com/10274858/61149474-29f08d80-a4d9-11e9-97d8-7c8105f8c945.png)
![our_image4](https://user-images.githubusercontent.com/10274858/61149475-29f08d80-a4d9-11e9-93ed-d72483e3abd7.png)

**Comment by quantoidb** [comment]
![our_image2](https://user-images.githubusercontent.com/10274858/61149942-31fcfd00-a4da-11e9-838e-13a3460deba2.png)

**Comment by quantoidb** [comment]
Image 2 

![our_image2](https://user-images.githubusercontent.com/10274858/61149998-4f31cb80-a4da-11e9-8aaa-e7e7ad2a005c.png)

**Comment by quantoidb** [comment]
image 3

![our_image3](https://user-images.githubusercontent.com/10274858/61150037-5ce75100-a4da-11e9-8fe3-c444e1d04be7.png)

**Comment by quantoidb** [comment]
image 4

![our_image4](https://user-images.githubusercontent.com/10274858/61150051-65d82280-a4da-11e9-9778-c5b99d724711.png)

**Comment by geojamesc** [comment]
Those are the pixels that fall inside the field boundary, as can be seen from this screenshot which shows the field boundaries in red outline against the pixels for one of the images you have listed in the r-studio console screenshot above:

![pixels_vs_field_boundary](https://user-images.githubusercontent.com/11922702/61200659-195b3580-a6da-11e9-9a0b-5b8e9c9e8b51.png)

This is a thin/narrow and elongated field and the S1 data has quite low spatial resolution (10m?) when compared to the size/detail of the field polygons. The algorithm within GDAL, the tool which I am using to clip the S1 data to the field boundaries has to make decisions when faced with pixels that fall partly within/outwith the field (polygon) boundary. 

 I seemed to recall there was some discusion of skipping such small, narow fields from the training data when the training data was being built from the zonal stats for the random forest. It sounds like the same could be done here, assuming the CNN is not going to use shape/size as one of it`s ways of classifying and that there is a note of which such fields were deemed to be to thin/narrow when you folks were doing the work for the random forest / zonal stats.

**Comment by quantoidb** [comment]
Nope, keras_model not accepting the crops images. 
I can plot and see them in R, but they are not feeding into the model for some reason. 

It's exactly the same code as my fossils project code, and that one runs like a dream. The only difference is fossils images are JPG and crop are .TIF
To accommodate .TIF I am changing from "gayscale" to "rgb", and from "3"  to "1" channels.

I have tried everything. Don't know where to go next.

**Comment by geojamesc** [comment]
Presumably the fossil images are normal 3 band RGB photographs provided as JPGs whereas the sentinel 1 images are a 2 band geotiff where each band (unlike the fossil image) is not a colour intensity but a backscatter value for a particular radar band. I've never used keras. When I did ml of images in the past from Python using OpenCV i.e. to build a KNN model, I read each image into a python numpy array and then passed to the classifier an array of arrays (the images) plus an array of the labels. Again I don't know anything about R/keras/machine-learning with R. I don`t know if input to keras can be an arbitrary array of data i.e. an array of S1 radar data or not.

**Comment by quantoidb** [comment]
Yes, that's exactly the difference between the images. I am digging into it more today. I've not given up :)

**Comment by quantoidb** [comment]
ok, I have a Python script to extract layer 1 (grayscale) image, so I can then feed it into NN model using R. This is just a stop gap for now to make sure the NN model working - it does. 

To run the model on Kelso, I need TIF images in separate folders by crop_type, to get this to work on the actual sample.

**Comment by quantoidb** [comment]
CORRECTION: To run the model on Kelso, I need the GRAYSCALE TIF images in separate folders by crop_type, to get this to work on the actual sample. 

James, I will explain next time we speak.

**Comment by quantoidb** [comment]
Could you do one more split please: Train 60%, Test 20%, Valid 20%. 

At this point it may be faster to split current Test in half and allocate between Test and Valid.

**Comment by quantoidb** [comment]
Could you do one more split please: Train 60%, Test 20%, Valid 20%.
At this point it may be faster to split current Test in half and allocate between Test and Valid.

In the end, I do not believe it will make any difference to the final result, but I want to be thorough. And we'll be ready for Scotland when that dataset is ready.

**Comment by quantoidb** [comment]
Hi @geojamesc 
Could you have a look at the attached PDF. I have 2 questions please: 

Q1 - Is it TRUE that if I look at a photo from: 
- BAND 1 - TRAIN folder - Class SPOT - photo 1, and 
- BAND 2 - TRAIN folder - Class SPOT - photo 1 
I would be looking at the same FIELD ? 

Q2 - Why is there a different TOTAL number of images in Band 1 and Band 2? 

Thanks!!
[Band1 and Band2 images and counts.pdf](https://github.com/cropmapteam/Scotland-crop-map/files/3473825/Band1.and.Band2.images.and.counts.pdf)

**Comment by quantoidb** [comment]
IGNORE Q2. Dumb Q. I'm not using all the classes so of course it won't add up to the same number 😂

**Comment by geojamesc** [comment]
The number after the "SpkRL" bit in the filename indicates the id of the field in question that the image describes i.e. 

S1B_20180630_30_asc_175813_175838_DV_Gamma0_GB_OSGB_RCTK_SpkRL_{fieldID}_resized.tif

so for:

S1B_20180630_30_asc_175813_175838_DV_Gamma0_GB_OSGB_RCTK_SpkRL_9_resized.tif

fieldId = 9

These should have persisted when the images were moved into their various sub-folders etc.

**Comment by quantoidb** [comment]
ok, so I took the first file from TrainB1, PC crop type, file name: S1A_20180102_132_asc_175018_175043_DV_Gamma-0_GB_OSGB_RCTK_SpkRL_37_resized.tif 

Then went to TrainB2, PC crop type, and took the file ending in _37_. They have exactly the same file names. And, when I plot it, look exactly the same too - see PDF. 

I did the same with image ending _75_. These also look exactly the same to me. 

What am I missing? 

[image_37_PC.pdf](https://github.com/cropmapteam/Scotland-crop-map/files/3476989/image_37_PC.pdf)
[image_75_PC.pdf](https://github.com/cropmapteam/Scotland-crop-map/files/3476990/image_75_PC.pdf)

**Comment by geojamesc** [comment]
The pixels present in each band should be the same since a pixel in band1 will occupy the same geographic space as a pixel in band2.

**Comment by quantoidb** [comment]
So if I plot Band 1 and Band 2 of the same image, to me, to the naked eye, they look exactly the same  - see attached PDF. 

How do I, or the algorithm, tell them apart then? I expected the shading at least to be somewhat different between the bands. 
I guess I still don't understand what the difference is between the bands. 

[image_75_PC.pdf](https://github.com/cropmapteam/Scotland-crop-map/files/3477083/image_75_PC.pdf)
[image_37_PC.pdf](https://github.com/cropmapteam/Scotland-crop-map/files/3477084/image_37_PC.pdf)

**Comment by geojamesc** [comment]
From what I understand the band1/2 - VV and VH are vertical and horizontal polarizations(?). From what I recall (though this recollection may not be correct since there was a lot of crop map going on at the time!), you extracted band2 from the clipped images as greyscaled images since when I ran your script on my own machine it produced blank images.

**Comment by quantoidb** [comment]
yes, I extract greyscale for the NN model. but here I am just plotting the images, and that's before I apply any "filetrs" to them. 
Maybe @spatial-scales can help. 

Richard, can you read my last comment/question above, and educate me on what an image with only Band1 would look like compared to the same image with only Band2 info.

**Comment by quantoidb** [comment]
I plotted a few images again. Here's the PDF. They are different. I think I have been sitting tied to the laptop for too long now :( 
[show_images-Band1_Band2.pdf](https://github.com/cropmapteam/Scotland-crop-map/files/3478343/show_images-Band1_Band2.pdf)
